File: csrc/free_blocks_ids.cpp
#include "free_blocks_ids.h"
#include <iostream>
#include <chrono>


bool compare_block_ptr(const Block* a, const Block* b) {
    return *b < *a;  // 假设 Block::operator< 是 const 方法
}
// ===== FreeBlockIdsBase =====
FreeBlockIdsBase::FreeBlockIdsBase(const std::vector<Block*>& blocks_,
                                   int offset_,
                                   const Config *config_,
                                   int minimum_scaling_block_count_)
    : config(config_), blocks(blocks_), offset(offset_),
      minimum_scaling_block_count(minimum_scaling_block_count_) {
    // std::make_heap(heap.begin(), heap.end(),compare_block_ptr);
}

void FreeBlockIdsBase::scale_up(int) {
    throw std::runtime_error("scale_up() not implemented");
}

void FreeBlockIdsBase::scale_down(int) {
    throw std::runtime_error("scale_down() not implemented");
}

void FreeBlockIdsBase::scale_up_with_specific_blocks(int) {
    throw std::runtime_error("scale_up_with_specific_blocks() not implemented");
}

void FreeBlockIdsBase::sync_num_free_blocks() {
    throw std::runtime_error("sync_num_free_blocks() not implemented");
}

void FreeBlockIdsBase::push(int block_id) {
    heap.push_back(blocks[block_id]);
    std::push_heap(heap.begin(), heap.end(),compare_block_ptr);
    // std::cout<<"堆顶："<<heap[0]->block_id<<std::endl;
}

void FreeBlockIdsBase::append(int block_id) { push(block_id); }

int FreeBlockIdsBase::pop() {
    if (heap.empty()) return -1;
    std::pop_heap(heap.begin(), heap.end(),compare_block_ptr);
    int id = heap.back()->block_id;  // ✅ 用智能指针直接访问对象成员
    heap.pop_back();                 // 智能指针出作用域会自动释放资源
    return id;
}

int FreeBlockIdsBase::peek() const {
    if (heap.empty()) return -1;
    return heap.front()->block_id;
}

bool FreeBlockIdsBase::remove(int block_id) {
    // 考虑 offset（保持原逻辑）
    block_id += offset;

    // 在 heap 中查找 block_id 匹配的元素
    auto it = std::find_if(heap.begin(), heap.end(),
        [block_id](Block* bptr) {
            return bptr && bptr->block_id == block_id; 
            // bptr && 避免解引用空指针
        });

    // 如果找到了
    if (it != heap.end()) {
        // 删除该元素（不释放内存，因为生命周期由外部统一管理）
        heap.erase(it);
        // 重新整理堆结构
        std::make_heap(heap.begin(), heap.end(), compare_block_ptr);
        // 返回 true 表示删除成功
        return true;
    }
    // 没找到
    return false;
}

void FreeBlockIdsBase::remove_batch_with_global_id(const std::vector<int>& global_block_ids) {

    std::set<int> idset(global_block_ids.begin(), global_block_ids.end());

    // 移除匹配的 Block*
    heap.erase(
        std::remove_if(heap.begin(), heap.end(),
            [&](Block* bptr) {
                return bptr && idset.count(bptr->block_id) > 0;
            }
        ),
        heap.end()
    );

    // 重新整理成堆
    std::make_heap(heap.begin(), heap.end(), compare_block_ptr);
}

void FreeBlockIdsBase::push_batch_with_global_id(const std::vector<int>& block_ids) {
    for (int id : block_ids) {
        heap.push_back(blocks[id - offset]); // blocks是Block*容器，所以直接push指针
    }
    std::make_heap(heap.begin(), heap.end(), compare_block_ptr); // 裸指针版本要带比较器
}

int FreeBlockIdsBase::num_blocks_lent() const {
    return static_cast<int>(blocks.size()) - static_cast<int>(heap.size());
}

int FreeBlockIdsBase::size() const { return heap.size(); }
int FreeBlockIdsBase::operator[](int index) const { return heap[index]->block_id; }

// ===== MasterFreeBlockIds =====
MasterFreeBlockIds::MasterFreeBlockIds(const std::vector<Block*>& blocks_,
                                       int offset_,
                                       const Config *config_,
                                       int minimum_scaling_block_count_)
    : FreeBlockIdsBase(blocks_, offset_, config_, minimum_scaling_block_count_),
      init_block_num(minimum_scaling_block_count_) {
    heap.assign(blocks.begin() + init_block_num, blocks.end());
    std::make_heap(heap.begin(), heap.end(), compare_block_ptr);
}

void MasterFreeBlockIds::scale_down(int n) {
    int num_blocks = minimum_scaling_block_count * n;
    std::vector<int> global_block_ids;
    for (int i = init_block_num; i < init_block_num + num_blocks; ++i) {
        global_block_ids.push_back(blocks[i]->block_id);
    }
    std::cout << "[Master] Global block IDs to remove: ";
    for (auto id : global_block_ids) std::cout << id << " ";
    std::cout << "\n";

    auto t1 = std::chrono::high_resolution_clock::now();
    remove_batch_with_global_id(global_block_ids);
    auto t2 = std::chrono::high_resolution_clock::now();
    std::cout << "[Master] scale_down took " << std::chrono::duration<double>(t2 - t1).count() << " sec\n";

    init_block_num += num_blocks;
}

void MasterFreeBlockIds::master_scale_down_many(int n) {
    int num_blocks = minimum_scaling_block_count * n;
    std::vector<int> global_block_ids;
    for (int i = init_block_num; i < init_block_num + num_blocks; ++i) {
        global_block_ids.push_back(blocks[i]->block_id);
    }
    std::cout << "即将被缩减的global_block_ids:";
    for (auto id : global_block_ids) std::cout << id << " ";
    std::cout << "\n";
    auto t1 = std::chrono::high_resolution_clock::now();
    remove_batch_with_global_id(global_block_ids);
    auto t2 = std::chrono::high_resolution_clock::now();
    std::cout << std::chrono::duration<double>(t2 - t1).count() << "\n";
    init_block_num += num_blocks;
}

// ===== SlaveFreeBlockIds =====
SlaveFreeBlockIds::SlaveFreeBlockIds(const std::vector<Block*>& blocks_,
                                     int offset_,
                                     const Config *config_,
                                     int minimum_scaling_block_count_)
    : FreeBlockIdsBase(blocks_, offset_, config_, minimum_scaling_block_count_),
      init_block_num(minimum_scaling_block_count_) {
    heap.assign(blocks.begin(), blocks.begin() + init_block_num);
    std::make_heap(heap.begin(), heap.end(), compare_block_ptr);
}

void SlaveFreeBlockIds::scale_up(int n) {
    int total_new = minimum_scaling_block_count * n;
    heap.insert(heap.end(),
        blocks.begin() + init_block_num,
        blocks.begin() + init_block_num + total_new);
    std::make_heap(heap.begin(), heap.end(), compare_block_ptr);

    init_block_num += total_new;
    std::cout << "[Slave] init_block_num after scale_up: " << init_block_num << "\n";
}

void SlaveFreeBlockIds::scale_up_with_specific_blocks(int num_blocks) {
    int n = (num_blocks + minimum_scaling_block_count - 1) / minimum_scaling_block_count;
    scale_up(n);
}

// ===== MultiFreeBlockIds =====
MultiFreeBlockIds::MultiFreeBlockIds(const std::vector<Block*>& blocks_,
                                     const std::vector<int>& num_blocks_start_end_,
                                     int local_num_blocks_,
                                     const Config *config_,
                                     const std::vector<int>& minimum_scaling_block_count_)
    : blocks(blocks_),
      num_blocks_start_end(num_blocks_start_end_),
      local_num_blocks(local_num_blocks_),
      config(config_),
      minimum_scaling_block_count(minimum_scaling_block_count_) 
{
    n_group = static_cast<int>(num_blocks_start_end.size()) - 1;
    n_blocks = static_cast<int>(blocks.size());
    n_free_blocks = n_blocks;

    // 外部组初始化
    for (int i = 0; i < n_group; ++i) {
        std::vector<Block*> seg;
        seg.reserve(num_blocks_start_end[i + 1] - num_blocks_start_end[i]); // 提前分配空间

        // 显式 push_back，不转移所有权
        for (auto it = blocks.begin() + num_blocks_start_end[i];
             it != blocks.begin() + num_blocks_start_end[i + 1];
             ++it) 
        {
            seg.push_back(*it); // 只是复制指针地址
        }

        multi_free_block_ids.push_back(
            new MasterFreeBlockIds(seg, num_blocks_start_end[i], config, minimum_scaling_block_count[i])
        );
    }

    // 本地组初始化
    std::vector<Block*> local_seg;
    local_seg.reserve(blocks.size() - num_blocks_start_end.back());

    for (auto it = blocks.begin() + num_blocks_start_end.back();
         it != blocks.end();
         ++it) 
    {
        local_seg.push_back(*it); // 裸指针复制
    }
    std::cout<<"初始化local_free_block_ids";
    for(int i=0;i<10;i++){
        std::cout<<local_seg[i]->block_id<<" ";
    }
    std::cout<<std::endl;
    local_free_block_ids = new MasterFreeBlockIds(local_seg, num_blocks_start_end.back(), config);

    counter = 0;

    std::cout << "n_free_blocks init:" << n_free_blocks << "\n";
    sync_num_free_blocks();
}



MultiFreeBlockIds::~MultiFreeBlockIds() {
    for (auto m : multi_free_block_ids) delete m;
    delete local_free_block_ids;
}

int MultiFreeBlockIds::get_group_idx(int block_id) const {
    for (int i = 0; i < n_group; ++i) {
        if (block_id >= num_blocks_start_end[i] && block_id < num_blocks_start_end[i + 1])
            return i;
    }
    return -1;
}

void MultiFreeBlockIds::append(int block_id) {
    n_free_blocks++;
    int group_id = get_group_idx(block_id);
    if (group_id != -1) {
        int block_id_in_group = block_id - num_blocks_start_end[group_id];
        multi_free_block_ids[group_id]->append(block_id_in_group);
    } else {
        int block_id_in_group = block_id - num_blocks_start_end.back();
        local_free_block_ids->append(block_id_in_group);
    }
}

void MultiFreeBlockIds::remove(int block_id) {
    n_free_blocks--;
    int group_id = get_group_idx(block_id);
    if (group_id != -1) {
        int block_id_in_group = block_id - num_blocks_start_end[group_id];
        multi_free_block_ids[group_id]->remove(block_id_in_group);
    } else {
        int block_id_in_group = block_id - num_blocks_start_end.back();
        local_free_block_ids->remove(block_id_in_group);
    }
}

int MultiFreeBlockIds::size() const { return n_free_blocks; }

int MultiFreeBlockIds::operator[](int index) {
    bool all_empty = true;
    for (auto g : multi_free_block_ids) {
        if (g->size() > 0) { all_empty = false; break; }
    }
    if (all_empty) {
        if (local_free_block_ids->size() == 0)
            throw std::runtime_error("external 和 local 的 block 都空了");
        return (*local_free_block_ids)[index];
    }

    for (int i = 0; i < n_group; ++i) {
        if (multi_free_block_ids[counter]->size() > 0) {
            break;
        }
        std::cout << "group_id:" << counter << " 的 block 已经全部用完\n";
        counter = (counter + 1) % n_group;
    }

    Block* external_block = blocks[multi_free_block_ids[counter]->operator[](index)];
    if (local_free_block_ids->size() == 0) {
        counter = (counter + 1) % n_group;
        return external_block->block_id;
    }

    Block* local_block = blocks[local_free_block_ids->operator[](index)];
// 调试打印
    // std::cout << "local_block:    hash=" << local_block->hash
    //         << "  block_id=" << local_block->block_id << std::endl;

    // std::cout << "external_block: hash=" << external_block->hash
    //       << "  block_id=" << external_block->block_id << std::endl;
    if (*external_block < *local_block) {
        counter = (counter + 1) % n_group;
        // std::cout<<"external_block:"<<external_block->block_id<<std::endl;
        return external_block->block_id;
    } else {
        // std::cout<<"local_block:"<<local_block->block_id<<std::endl;
        return local_block->block_id;
    }
}
int MultiFreeBlockIds::peek() {
    // int block_id = (*this)[0];
    return (*this)[0];
}

void MultiFreeBlockIds::master_scale_down_many(int slave_idx, int n) {
    multi_free_block_ids[slave_idx]->master_scale_down_many(n);
    int num_blocks = minimum_scaling_block_count[slave_idx] * n;
    n_free_blocks -= num_blocks;
}

void MultiFreeBlockIds::sync_num_free_blocks() {
    n_free_blocks = 0;
    for (auto f : multi_free_block_ids) {
        n_free_blocks += f->size();
    }
    n_free_blocks += local_free_block_ids->size();
    // std::cout << "n_free_blocks sync:" << n_free_blocks << "\n";
}
// void MultiFreeBlockIds::hello_world() {
//     std::cout << "hello world \n";
// }

File: swiftcache/engine/block_manager/free_blocks_ids.py
import heapq
import time
from typing import Optional
from swiftcache.engine.block_manager.block import Block
from swiftcache.config import Config
class FreeBlockIdsBase:
    def __init__(self, blocks, offset: int = 0, config=None, minimum_scaling_block_count=0):
        self.config = config
        self.blocks = blocks
        self.offset = offset
        self.minimum_scaling_block_count = minimum_scaling_block_count
        self._heap = []

    # ======== Interface-like methods ========
    def scale_up(self, *args, **kwargs):
        """Expand resources (should be overridden by subclass)"""
        raise NotImplementedError("scale_up() is not implemented in this class.")

    def scale_down(self, *args, **kwargs):
        """Shrink resources (should be overridden by subclass)"""
        raise NotImplementedError("scale_down() is not implemented in this class.")

    # ======== Shared methods ========
    def push(self, block_id: int):
        block = self.blocks[block_id]
        heapq.heappush(self._heap, block)

    def append(self, block_id: int):
        self.push(block_id)

    def pop(self) -> Optional[int]:
        if not self._heap:
            return None
        block = heapq.heappop(self._heap)
        return block.block_id

    def peek(self) -> Optional[int]:
        if not self._heap:
            return None
        return self._heap[0].block_id
    
    def remove(self, block_id: int) -> bool:
        """根据 block_id 删除堆中对应的 Block"""
        block_id += self.offset
        for i, block in enumerate(self._heap):
            if block.block_id == block_id:
                # 删除该元素
                # print(f"即将删除{block_id}")
                self._heap.pop(i)
                # 重新建堆
                heapq.heapify(self._heap)
                return True
        return False  # 没找到

    def remove_batch_with_global_id(self, global_block_ids: list[int]):
        global_block_ids_set = set(global_block_ids)
        self._heap = [block for block in self._heap if block.block_id not in global_block_ids_set]
        heapq.heapify(self._heap)

    def push_batch_with_global_id(self, block_ids: list[int]):
        new_blocks = [self.blocks[b_id - self.offset] for b_id in block_ids]
        self._heap.extend(new_blocks)
        heapq.heapify(self._heap)

    @property
    def num_blocks_lent(self):
        return len(self.blocks) - len(self._heap)

    def __len__(self):
        return len(self._heap)

    def __getitem__(self, index: int):
        return self._heap[index].block_id


class MasterFreeBlockIds(FreeBlockIdsBase):
    def __init__(self, blocks, offset=0, config=None, minimum_scaling_block_count=0):
        super().__init__(blocks, offset, config, minimum_scaling_block_count)
        # Master keeps all blocks except initial reserved ones
        self.init_block_num = minimum_scaling_block_count
        self._heap = list(blocks[self.init_block_num:])
        heapq.heapify(self._heap)

    def scale_down(self, n: int):
        """Master-specific shrink"""
        num_blocks = self.minimum_scaling_block_count * n
        global_block_ids = [
            block.block_id for block in
            self.blocks[self.init_block_num:self.init_block_num + num_blocks]
        ]
        print(f"[Master] Global block IDs to remove: {global_block_ids}")
        t = time.perf_counter()
        self.remove_batch_with_global_id(global_block_ids)
        print(f"[Master] scale_down took {time.perf_counter() - t:.6f} sec")
        self.init_block_num += num_blocks

    def master_scale_down_many(self, n:int):
        num_blocks = self.minimum_scaling_block_count * n
        global_block_ids = [block.block_id for block in self.blocks[self.init_block_num: self.init_block_num + num_blocks]]
        print(f'即将被缩减的global_block_ids:{global_block_ids}')
        t = time.perf_counter()
        self.remove_batch_with_global_id(global_block_ids)
        print(time.perf_counter()-t)
        self.init_block_num += num_blocks


class SlaveFreeBlockIds(FreeBlockIdsBase):
    def __init__(self, blocks, offset=0, config=None, minimum_scaling_block_count = 1):
        super().__init__(blocks, offset, config, minimum_scaling_block_count)
        # Slave gets initial N blocks
        self.init_block_num = minimum_scaling_block_count
        self._heap = list(blocks[:self.init_block_num])
        heapq.heapify(self._heap)

    def scale_up(self, n: int = 1):
        """Slave-specific expansion"""
        self._heap.extend(
            self.blocks[self.init_block_num:
                        self.init_block_num + self.minimum_scaling_block_count * n]
        )
        heapq.heapify(self._heap)
        self.init_block_num += self.minimum_scaling_block_count * n
        print(f"[Slave] init_block_num after scale_up: {self.init_block_num}")
    
    def scale_up_with_specific_blocks(self,num_blocks:int):
        n = (num_blocks + self.minimum_scaling_block_count - 1) // self.minimum_scaling_block_count
        self.scale_up(n)


class MultiFreeBlockIds:
    def __init__(self,blocks: list[Block], num_blocks_start_end: list[int],local_num_blocks,config:Config,minimum_scaling_block_count:int):
        # assert len(blocks) == num_blocks_start_end[-1]
        self.blocks = blocks
        self.minimum_scaling_block_count = minimum_scaling_block_count
        self.n_group = len(num_blocks_start_end)-1
        self.num_blocks_start_end = num_blocks_start_end
        self.local_num_blocks = len(blocks) - num_blocks_start_end[-1]
        self.n_blocks = len(blocks)
        self.multi_free_block_ids = []
        self.n_free_blocks = self.n_blocks
        for i in range(self.n_group):
            free_blocks_ids = MasterFreeBlockIds(blocks[num_blocks_start_end[i]:num_blocks_start_end[i+1]],num_blocks_start_end[i],config,minimum_scaling_block_count[i])
            self.multi_free_block_ids.append(free_blocks_ids)
        self.local_free_block_ids = MasterFreeBlockIds(blocks[num_blocks_start_end[-1]:], num_blocks_start_end[-1], config)
        self.counter = 0
        print(f'n_free_blocks init:{self.n_free_blocks}')
        self.sync_num_free_blocks()
    def get_group_idx(self, block_id):
        for i in range(self.n_group):
            if block_id >= self.num_blocks_start_end[i] and block_id < self.num_blocks_start_end[i+1]:
                return i
        
        return -1
        

    def append(self,block_id):
        self.n_free_blocks += 1
        group_id = self.get_group_idx(block_id)
        block_id_in_group = block_id - self.num_blocks_start_end[group_id] # -offset
        if group_id != -1:
            self.multi_free_block_ids[group_id].append(block_id_in_group)
        else:
            self.local_free_block_ids.append(block_id_in_group)

    def remove(self, block_id):
        self.n_free_blocks -= 1
        group_id = self.get_group_idx(block_id)
        block_id_in_group = block_id - self.num_blocks_start_end[group_id] # -offset
        if group_id != -1:
            self.multi_free_block_ids[group_id].remove(block_id_in_group)
        else:
            self.local_free_block_ids.remove(block_id_in_group)

    def __len__(self):
        return self.n_free_blocks

    def __getitem__(self, index: int):
        # 如果所有 external group 都空，直接处理
        if all(len(g) == 0 for g in self.multi_free_block_ids):
            if len(self.local_free_block_ids) == 0:
                raise ValueError("external 和 local 的 block 都空了")
            # external 都空 → 返回 local
            return self.local_free_block_ids._heap[index].block_id

        # 找到第一个非空的 external group
        for _ in range(self.n_group):
            if len(self.multi_free_block_ids[self.counter]) > 0:
                break
            print(f"group_id:{self.counter} 的 block 已经全部用完")
            self.counter = (self.counter + 1) % self.n_group

        external_block = self.multi_free_block_ids[self.counter]._heap[index]

        # 如果 local 为空，直接返回 external
        if len(self.local_free_block_ids) == 0:
            return external_block.block_id

        local_block = self.local_free_block_ids._heap[index]
        # print(f'{external_block.block_id},{local_block.block_id}')
        if external_block < local_block:
            self.counter = (self.counter + 1) % self.n_group
            return external_block.block_id
        else:
            return local_block.block_id
    
    def master_scale_down_many(self,slave_idx:int, n:int):
        self.multi_free_block_ids[slave_idx].master_scale_down_many(n)
        num_blocks = self.minimum_scaling_block_count[slave_idx] * n
        self.n_free_blocks - num_blocks
    
    def sync_num_free_blocks(self):
        self.n_free_blocks = 0
        for free_blocks_ids  in self.multi_free_block_ids:
            self.n_free_blocks += len(free_blocks_ids)
        self.n_free_blocks += len(self.local_free_block_ids)
        # print(f'n_free_blocks sync:{self.n_free_blocks}')